### ISSUE TITLE
how does versioning work? get the latest version of vips?

### ISSUE THREAD

**Issue by jrochkind** [heroku/heroku-buildpack-apt#82]
We are interested in the `vips` package. But we discovered we actually need the latest minor version of vips, 8.10.x.  8.9.x has problems for us, we need the latest. 

When we just add `vips` to Aptfile, we get an 8.9 version. (We are running on heroku-20 using ubuntu 20, not sure if that matters for what version of thing you will get from Aptfile). 

In an attempt to force the particular more recent version of vips we want, we add to Aptfile a specfic url:

    http://ftp.us.debian.org/debian/pool/main/v/vips/libvips-tools_8.10.5-1_amd64.deb 

But then we get errors like: `LoadError: Could not open library 'vips.so.42': vips.so.42: cannot open shared object file: No such file or directory.`

Not sure what's going on with that.

I realize I don't have a handle on the basic theory/model here, not being super familiar with sysadmin'ing linux packages from the start (hey that's why we're on heroku!).  What determines what version gets installed when I just add `vips` to Aptfile? Ought I be able to force an 8.10 version, is it expected there should be a way to do that? And if so, through what manner?

**Comment by edmorley** [comment]
Hi! This buildpack has to use a slightly hacky way of installing packages, since sudo is not available either at build time or run time, and as such any paths outside of `/app` or `/tmp` are read only.

The buildpack:
1. Copies the stack image's base APT sources list into the build cache (which for Heroku-20 comes from [here](https://github.com/heroku/stack-images/blob/2a93b01a25a512cdae9d1d8db76e39e3f94c1dfe/heroku-20/setup.sh#L12-L28))
2. Adds any custom APT repository sources specified via the `:repo:` syntax in `Aptfile` (see README for example)
3. Does an `apt-get update` using the resultant APT sources list
4. For any package URLs in `Aptfile`, curls the file and stores in the cache
5. For any packages in `Aptfile` that aren't URLs, does an `apt-get install` with the `-d` option, which just downloads the archives
6. For all downloaded .deb archives, extracts them manually into the build directory using `dpkg -x`
7. Sets up a bunch of environment variables so these packages can *hopefully* be found the same as if they were installed in the standard global system locations. 

The implications of this are:
* Unless custom APT repos are added or .deb URLs used, packages must exist for the respective Ubuntu LTS on https://packages.ubuntu.com
* No dependency resolution occurs, since the packages are naively downloaded/extracted by name. Ie: Any sub-dependencies of packages must be added explicitly to `Aptfile` otherwise you may get missing library errors.
* Any packages that do things in their post install scripts might not work correctly, since those scripts don't get run when using `dpkg -x` (extracting the archive only).
* Any package version specifier supported by apt-get, should in theory be supported as a line in `Aptfile`, since the buildpack just passes those entries along to `apt-get`.

Checking https://packages.ubuntu.com I see only Ubuntu 20.10+ (groovy) has VIPS 8.10.x, with 20.04 (focal) still being on 8.9.x:
https://packages.ubuntu.com/focal/libvips-dev
https://packages.ubuntu.com/groovy/libvips-dev

This isn't too surprising, since anything but bug fixes and security updates typically isn't backported to an LTS once it's released, and libvps 8.10 was released 2 months after Ubuntu 20.04:
https://libvips.github.io/libvips/2020/06/18/What's-new-in-8.10.html

I was about to write more, but I've just spotted this buildpack which might have solved the problem already for you:
https://github.com/brandoncc/heroku-buildpack-vips-h20

**Comment by jrochkind** [comment]
Thank you that is all super helpful @Edmorley. 

Wow, this stuff gets complicated. 

We think vips 8.10 does fix a bug we were seeing in vips 8.9, but I realize that doesn't mean it'll necessarily get backported (on what timeframe) to ubuntu LTS... I wasn't even sure how to check what version of what dependency is in LTS for what ubuntu version. 

Ubuntu 20.10 but not 20.04, eh? I don't totally understand what *that* means, but since all the heroku stacks are ubuntu *.04 versions, ah, those are the LTS versions? So basically, not gonna get an ubuntu with vips 8.10+ for a *while* on ubuntu lts/heroku, if I understand it right. 

Yeah, we are using a specificcally vips buildpack, which does seem to be working. That's another confusing issue, keeping track of all the third-party buildpacks, it seems to be common practice for them to be stuck to a *particular* dependency version, and everyone to just fork it on their own if they want a different version, so we end up with like dozens of "vips" buildpacks listed in the heroku directory, with it being unclear what version any given one will install, or if they are currently actually used by anyone let alone supported. 

So many moving parts!

But this was a helpful overview thanks.

**Comment by jrochkind** [comment]
@edmorley hmm, if I understand the implications of what you're saying (and I may not), it seems like while the README has an example:

```
# or include links to specific .deb files
https://downloads.example.com/example.deb
```

This is actually very unlikely to work? Maybe unless you have a carefully constructed .deb file specifically intended for this environment? 

No?

**Comment by jrochkind** [comment]
And, 

```
Any package version specifier supported by apt-get, should in theory be supported as a line in Aptfile, since the buildpack just passes those entries along to apt-get.
```

Not sure what that looks like, should an example of *that* be given in README? (Or is it unlikely to be very helpful anyway? I understand in this example it would not have been)

**Comment by edmorley** [comment]
> We think vips 8.10 does fix a bug we were seeing in vips 8.9, but I realize that doesn't mean it'll necessarily get backported (on what timeframe) to ubuntu LTS

It might be worth filing a bug at https://launchpad.net/ubuntu/+source/vips/+bugs with links to the upstream bug fix requesting a backport to Ubuntu 20.04.

> Ubuntu 20.10 but not 20.04, eh? I don't totally understand what _that_ means, but since all the heroku stacks are ubuntu *.04 versions, ah, those are the LTS versions?

Ubuntu releases are named `<Year>.<Month>` of their release. So Ubuntu 20.04 was the April 2020 release etc.
Ubuntu releases new versions every 6 months. Every 2 years one of those releases is a special "Long Term Support" release (aka LTS).
Those LTS releases are what the majority of servers/infrastructure providers/... will run instead of the 6 monthly releases.

There's more on this at:
https://ubuntu.com/about/release-cycle

> it seems like while the README has an example:
> This is actually very unlikely to work? Maybe unless you have a carefully constructed .deb file specifically intended for this environment?

So it really depends on the package. If it's something that depend on lots of other system libraries and was compiled for a completely different version on Ubuntu (or even a different Linux distro), then yes it might not work.

However there are lots of third party packages out there that are compatible with any "debian/ubuntu-like" distro and will install just fine.

In the case of your failure with the URL, I suspect the issue was more that `libvips-tools` depends on `libvps` so you really needed multiple URLs, one for each of the packages in the dependency tree.

> Not sure what that looks like, should an example of that be given in README? (Or is it unlikely to be very helpful anyway? I understand in this example it would not have been)

So in general the distro will only make one minor version of a package available, albeit with several different patch versions. So I suspect there aren't many times when it would be useful.
